actioncable/app/assets/javascripts/action_cable/connection.coffee

```diff
@@ -108,15 +110,17 @@ class ActionCable.Connection
       ActionCable.log("WebSocket onclose event")
       if not @isProtocolSupported()
         ActionCable.log("WebSocket is closing because there is no common protocol. Stopping ConnectionMonitor.")
-        @monitor.stop()   
-      @disconnect()
+        @disconnect(ActionCable.Events.Disconnected.ProtocolNotSupported)
+      else
+        @disconnect(new ActionCable.Events.Disconnected(event.code, event.reason, true))
 
     error: ->
       ActionCable.log("WebSocket onerror event")
-      @disconnect()
 
-  disconnect: ->
-    return if @disconnected
+  disconnect: (event) ->
+    if event && !event.isRecoverable
+      @monitor.stop()
```

Comment:
Think this should move to the `close` handler so it's more closely tied to the related condition.